decoder/internal/walker/walker.go

```diff
@@ -24,14 +24,23 @@ type Walker interface {
 func Walk(ctx context.Context, node hclsyntax.Node, nodeSchema schema.Schema, w Walker) hcl.Diagnostics {
 	var diags hcl.Diagnostics
 
+	blkNestingLvl, ok := schemacontext.BlockNestingLevel(ctx)
+	if !ok {
+		ctx = schemacontext.WithBlockNestingLevel(ctx, 0)
+		blkNestingLvl = 0
+	}
+
 	switch nodeType := node.(type) {
 	case *hclsyntax.Body:
+		bodyCtx := ctx
 		foundBlocks := make(map[string]uint64)
 		dynamicBlocks := make(map[string]uint64)
-		bodySchema, ok := nodeSchema.(*schema.BodySchema)
-		if ok {
```

Comment:
Why are we getting rid of this check here? I see you moved it to the two for loops for attribute and blocks.
With this change, we call walk in lines 60 and 87, even without schemas.

Reply:
This was done in a separate commit - which I hoped would provide the context/clarity 😅 https://github.com/hashicorp/hcl-lang/pull/341/commits/6d24e31a06cd05f1c058dcf1d7c38ac82798467b

The point is that we previously skipped walking body for which we don't have schema. Now this is _probably_ fine for all the existing use cases we have for the validation as we cannot really do much validation without schema anyway and all the validators short-circuit on lack of schema early on. However, this is inconsistent with how we treat other AST nodes (blocks or attributes), so I made it consistent here.

It also helped in the walker tests where I did not have to supply any schema just to verify that walking works.

Reply:
I see, that makes sense. But do we want to report diagnostics for every attribute/block within a block without a schema? Feels nosy to me.

**Before**
<img width="694" alt="CleanShot 2023-11-07 at 13 26 51@2x" src="https://github.com/hashicorp/hcl-lang/assets/45985/e0737516-dd22-433c-8fd8-1162bbcb8aa5">

**After**
<img width="706" alt="CleanShot 2023-11-07 at 13 24 24@2x" src="https://github.com/hashicorp/hcl-lang/assets/45985/ea502087-9584-43b7-b547-e35dc5be58b0">

Reply:
Good question. I _think_ we could suppress that mentioned noise by adding `ctx = schemacontext.WithUnknownSchema(ctx)` - equivalent to when merging fails?

I can add that patch.

Reply:
done, PTAL